torchdata/datapipes/map/util/cacheholder.py

```diff
@@ -0,0 +1,54 @@
+# Copyright (c) Facebook, Inc. and its affiliates.
+from typing import Callable, List, Optional, TypeVar
+
+from torchdata.datapipes import functional_datapipe
+from torchdata.datapipes.map import MapDataPipe
+
+
+T_co = TypeVar("T_co", covariant=True)
+
+
+def list_cache(size: int) -> List:
+    cache = [None] * size
+    try:
+        from numpy import array
+
+        cache = array(cache)
+    except (ImportError, ModuleNotFoundError):
+        pass
+    return cache
+
+
+@functional_datapipe("in_memory_cache")
+class InMemoryCacheHolderMapDataPipe(MapDataPipe[T_co]):
+    r"""
+    Stores elements from the source DataPipe in memory (functional name: ``in_memory_cache``).
+
+    Args:
+        source_dp: source DataPipe from which elements are read and stored in memory
+        cache_fn: function that takes in an integer size and return a cache object that implements ``__getitem__``
+
+    Example:
+        >>> from torchdata.datapipes.map import SequenceWrapper
+        >>> source_dp = SequenceWrapper(range(10))
+        >>> cache_dp = source_dp.in_memory_cache()
+        >>> list(cache_dp)
+        [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
+    """
+
+    def __init__(self, source_dp: MapDataPipe[T_co], cache_fn: Optional[Callable] = None) -> None:
+        self.source_dp: MapDataPipe[T_co] = source_dp
```

Comment:
If the source_dp is a `MapDataPipe` and it holds the memory of the Dataset, why we need this DataPipe as all data has been stored in `source_dp`.

Reply:
1. IIUC, since `MapDataPipe` is lazily computed, the operations are applied every time `__getitem__` is called.

Here is a silly example:
```python
def map_dp_quick_test(cached: bool):
    print(f"cached={cached}")
    dp = SequenceWrapper(range(100000))
    dp = dp.map(lambda x: x + 1)
    dp = dp.map(lambda x: x * 10)
    dp = dp.map(lambda x: x + 1)
    dp = dp.map(lambda x: x * 10)
    dp = dp.map(lambda x: x + 1)
    if cached:
        dp = dp.in_memory_cache()
    dps = [list(dp.map(lambda x: x + 1)) for _ in range(100)]

map_dp_quick_test(cached=True)
gc.collect()
map_dp_quick_test(cached=False)
```
Results:
```
[4.35055778s] map_dp_quick_test(cached=True)
[14.84188485s] map_dp_quick_test(cached=False)
```

It may be useful to cache these operations if `cache_dp` ended up being used by multiple DataPipes downstream. It is unlikely to be as useful as the `IterDataPipe` version though.

2. I think it will be helpful to provide this as a pattern to encourage using `numpy.array` (or similar wrapper) to avoid the "copy on write" memory issue that some users have experienced.

3. Lastly, I have seen some questions by users about caching between epoch while having multiple workers. I think using `multiprocessing.Array` here can achieve that goal (I am not sure if that should be the recommended pattern, but it is a possibility). Let me know what you think.

Reply:
> > 1. IIUC, since `MapDataPipe` is lazily computed, the operations are applied every time `__getitem__` is called.
> 
> It makes sense
> 
> For 2 and 3, we do want to support such behavior. But, I don't know this is the right approach. For 2, we are adding a dependency of numpy. For 3, the overhead of `mp.Array` is really high IIRC
> 
> I would love to use list as the default data structure but let user can provide different class for cache if they want to take care of COW problem

Do you think the soft dependency on `numpy` as implemented here is reasonable? I made that choice because I feel like users aren't usually aware of the COW problem until it is a very big one. Using `numpy.array` by default here somewhat mitigates that problem (or at least prompt user to take a look if we describe it well in the docstring).

I don't plan to use `mp.Array` by default here but it will be good to mention it and provide an example.